tests/unit/test_analyzer_util_suite.py

```diff
@@ -428,5 +441,229 @@ def test_create_dynamic_iceberg_table():
-        " = 'OPTIMIZED' AS  SELECT  *  FROM (select * from foo)"
+        " = 'OPTIMIZED' AS  SELECT  * \n"
+        " FROM (\nselect * from foo\n)"
+    )
+
+
+def test_project_statement_formatting():
+    print(project_statement(["col1", "col2"], "table1"))
+    assert project_statement(["col1", "col2"], "table1") == (
+        " SELECT \n" "    col1, \n" "    col2\n" " FROM (\n" "table1\n" ")"
+    )
+
+    assert project_statement(["col1 as a", "col2 as b"], "table1") == (
+        " SELECT \n" "    col1 as a, \n" "    col2 as b\n" " FROM (\n" "table1\n" ")"
+    )
+
+    assert project_statement(
+        ["CASE WHEN col1 > 0 THEN 1 ELSE 0 END as flag", "COUNT(*) as cnt"], "table1"
+    ) == (
+        " SELECT \n"
+        "    CASE WHEN col1 > 0 THEN 1 ELSE 0 END as flag, \n"
+        "    COUNT(*) as cnt\n"
+        " FROM (\n"
+        "table1\n"
+        ")"
+    )
+
+    child_query = "SELECT a, b\nFROM table1\nWHERE x > 0"
+    assert project_statement(["col1", "col2"], child_query) == (
+        " SELECT \n"
+        "    col1, \n"
+        "    col2\n"
+        " FROM (\n"
+        "SELECT a, b\n"
+        "FROM table1\n"
+        "WHERE x > 0\n"
+        ")"
+    )
+
+    print(project_statement([], "table1"))
+    assert project_statement([], "table1") == (
+        " SELECT  * \n" " FROM (\n" "table1\n" ")"
+    )
+
+    assert project_statement(["col1", "col2"], "table1", is_distinct=True) == (
+        " SELECT  DISTINCT \n" "    col1, \n" "    col2\n" " FROM (\n" "table1\n" ")"
+    )
+
+
+def test_nested_query_formatting():
+    nested_query = project_statement(
+        ["t.col1", "t.col2"],
+        project_statement(["inner.a as col1", "inner.b as col2"], "base_table inner"),
+    )
+    assert nested_query == (
+        " SELECT \n"
+        "    t.col1, \n"
+        "    t.col2\n"
+        " FROM (\n"
+        " SELECT \n"
+        "    inner.a as col1, \n"
+        "    inner.b as col2\n"
+        " FROM (\n"
+        "base_table inner\n"
+        ")\n"
+        ")"
+    )
+
+
+def test_table_function_statement_formatting():
+    assert table_function_statement("my_table_func()") == (
+        " SELECT  * \n" " FROM (\n" " TABLE (my_table_func())\n" ")"
+    )
+
+    assert table_function_statement("my_table_func()", ["col1", "col2"]) == (
+        " SELECT \n"
+        "    col1, \n"
+        "    col2\n"
+        " FROM (\n"
+        " TABLE (my_table_func())\n"
+        ")"
+    )
+
+
+def test_filter_statement_formatting():
+    assert filter_statement("x > 0 AND y < 10", "my_table") == (
+        " SELECT  * \n" " FROM (\n" "my_table\n" ")\n" " WHERE x > 0 AND y < 10"
+    )
+
+
+def test_sample_by_statement_formatting():
+    # We use regex to match the output because the percent rank col is randomly generated
+    sample = sample_by_statement(
+        child="my_table", col="category", fractions={"A": 0.1, "B": 0.5, "C": 1.0}
+    )
+    print(sample)
+    expected_pattern = re.compile(
```

Comment:
print should be removed from here. Also another tip for this case, we can use unittest.mock.patch to patch the return value of random column generator name to be deterministic for this test. In that case, we can do exact string matching instead of regex matching. 

https://github.com/snowflakedb/snowpark-python/blob/44f0db743b5bf2181ecbbcee8ce65d15481b9e00/tests/unit/test_udtf.py#L31